hypervisor/qemu/qemu_machine.go

```diff
@@ -0,0 +1,216 @@
+package qemu
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type State int
+
+const (
+	STOPPED State = iota // 0
+	STARTING
+	RUNNING
+	STOPPING
+	REBOOTING
+	SHUTTINGDOWN
+	TERMINATING
+	FAILED
+)
+
+var stateValues = map[State]string{
+	STOPPED:      "STOPPED",
+	STARTING:     "STARTING",
+	RUNNING:      "RUNNING",
+	STOPPING:     "STOPPING",
+	REBOOTING:    "REBOOTING",
+	SHUTTINGDOWN: "SHUTTINGDOWN",
+	FAILED:       "FAILED",
+}
+
+var MachineState = map[string]State{
+	"STOPPED":      STOPPED,
+	"STARTING":     STARTING,
+	"RUNNING":      RUNNING,
+	"STOPPING":     STOPPING,
+	"REBOOTING":    REBOOTING,
+	"SHUTTINGDOWN": SHUTTINGDOWN,
+	"TERMINATING":  TERMINATING,
+	"FAILED":       FAILED,
+}
+
+type Machine struct {
+	State             State
+	Cores             int
+	Memory            uint64
+	Name              string
+	Display           string
+	Vnc               string
+	MonitorSocketPath string
+	SerialSocketPath  string
+	Pidfile           string
+	Nics              []NetDev
+	Drives            map[string]Drive
+	Process           *os.Process
+	Kvm               bool
+}
+
+type NetDev struct {
+	IfName       string
+	Index        string
+	Ipv4Addr     string
+	MacAddr      string
+	Bridge       string
+	BridgeHelper string
+	Type         string
+}
+
+func (m *Machine) promptPattern() string {
+	return fmt.Sprintf("openvdc@%s", m.Name)
+}
+
+func (m *Machine) HavePrompt() bool {
+	c, err := net.Dial("unix", m.SerialSocketPath)
+	if err != nil {
+		return false
+	}
+
+	buf := bufio.NewReader(c)
+	matchprompt := make(chan bool, 1)
+	defer close(matchprompt)
+
+	go func() {
+		defer c.Close()
+		c.SetReadDeadline(time.Now().Add(time.Second))
+		tries := 0
+		for {
+			if tries > 10 {
+				matchprompt <- false
+				return
+			}
+			line, _, _ := buf.ReadLine()
+			if strings.Contains(string(line), m.promptPattern()) {
+				matchprompt <- true
+				return
+			}
+			tries = tries + 1
+		}
+	}()
+
+	// send new line in order to trigger the prompt
+	fmt.Fprintf(c, "\n")
+	return <-matchprompt
+}
+
+func (m *Machine) WaitForPrompt() bool {
+	c, err := net.Dial("unix", m.SerialSocketPath)
+	defer c.Close()
+	if err != nil {
+		return false
+	}
+	buf := bufio.NewReader(c)
+
+	if err := c.SetReadDeadline(time.Now().Add(5 * time.Second)); err != nil {
+		return false
+	}
+	b, _ := buf.ReadBytes('\n')
+	return (strings.Contains(string(b), m.promptPattern()))
+}
+
+// since machine struct does not get saved in memory for each instance there may not be any points
+// in scheduling states as they are not stored anywhere
+func (m *Machine) ScheduleState(nextState State, timeout time.Duration, evaluation func() bool) error {
+	errorc := make(chan error)
```

Comment:
Good to know that go channel is GCed. From my experience, the closing explicitly helps to find logical issue of channels because the writing or reading to the closed channel causes panic.
  
It is easy to write the code that causes leaking goroutine and channel.

https://stackoverflow.com/questions/38090545/are-goroutines-garbage-collected-together-with-their-channels